### PR TITLE
feat: merge feature sets

### LINKED ISSUE #277: Add option to specify multiple datasets to concatenate before training
Allows us to incremental feature sets that we can merge/concat directly instead of hacking something together. 
Remember not to replace keys.

Implementation:

- [ ] Hash uuid_col
- [ ] Check if exact match between the two --> concat, else join
- [ ] If different amount of rows, raise error

### PR DESCRIPTION
**Pull request by HLasse**
If one specifies multiple paths in `DataSchema.dir` the dataloader will now merge them. This allows us to build upon datasets with by incrementally adding features (without too much hacking). Fixes #277

### CODE REVIEW

File: psycop/common/model_training/config_schemas/data.py [hunk 1]
```diff
@@ -27,7 +27,7 @@ class ColumnNamesSchema(PSYCOPBaseModel):
 class DataSchema(PSYCOPBaseModel):
     """Data configuration."""
 
-    dir: Union[Path, str]  # Location of the dataset # noqa
+    dir: list[Path | str] | Path | str  # Location of the dataset # noqa
```
Comment (MartinBernstorff) [inline psycop/common/model_training/config_schemas/data.py:30]:
```suggestion
    dir: list[DirectoryPath] | Path  # Location of the dataset # noqa
```

https://docs.pydantic.dev/1.10/usage/types/#pydantic-types

File: psycop/common/model_training/data_loader/data_loader.py [hunk 2]
```diff
@@ -42,16 +42,97 @@ def _check_column_names(self, df: pd.DataFrame):
...
+
+    @staticmethod
+    def _check_dataframes_can_be_joined(
+        datasets: list[pd.DataFrame],
+        uuid_column: str,
+    ) -> bool:
+        """Check if pred_time_uuid columns contain the same elements so they can
+        be joined without introducing new rows"""
+        base_uuid = datasets[0][uuid_column]
+        return all(base_uuid.isin(df[uuid_column]).all() for df in datasets[1:])
```
Comment (MartinBernstorff) [inline psycop/common/model_training/data_loader/data_loader.py:63]: I have a very hard time parsing this; we should probably split it out over multiple lines for readability.

File: psycop/common/model_training/data_loader/data_loader.py [hunk 3]
```diff
@@ -42,16 +42,97 @@ def _check_column_names(self, df: pd.DataFrame):
...
+        n_rows = [dataset.shape[0] for dataset in datasets]
+        if len(set(n_rows)) != 1:
+            raise ValueError(
+                "The datasets have a different amount of rows. "
+                + "Ensure that they have been created with the same "
+                + "prediction times.",
+            )
+        shared_id_columns = [
+            self.data_cfg.col_name.id,
+            self.data_cfg.col_name.pred_time_uuid,
```
Comment (MartinBernstorff) [inline psycop/common/model_training/data_loader/data_loader.py:90]: This col is the concate of the id and pred_timestamp cols, so we only need to check that one :-)

**Review by MartinBernstorff — COMMENTED** [review]
Super nice work, Lasse! Only a handful of comments.

File: psycop/common/model_training/data_loader/data_loader.py (hunk 3 above)
Comment (HLasse) [reply psycop/common/model_training/data_loader/data_loader.py:90]: Only checking that one (see `_check_dataframes_can_be_concatenated`) - these are the columns that need to be dropped for concatenation :)

File: psycop/common/model_training/data_loader/tests/test_feature_merging.py [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+
+def test_check_dataframes_can_be_concatenated_false(
+    base_feature_df: pd.DataFrame,
+    feature_df_different_order_uuids: pd.DataFrame,
+):
+    with pytest.raises(AssertionError):
+        assert DataLoader._check_dataframes_can_be_concatenated(
+            datasets=[base_feature_df, feature_df_different_order_uuids],
+            uuid_column="prediction_time_uuid",
+        )
```
[the remark this answers is not in the document]
Comment (HLasse) [reply psycop/common/model_training/data_loader/tests/test_feature_merging.py:77]: <3 Is there a nice way to parameterise when one of them has to raise an assertion?

[the remark this answers is not in the document]
Comment (HLasse) [reply psycop/common/model_training/data_loader/tests/test_feature_merging.py:77]: lol gg test

**Pull request merged**